Allow `align_to` to take in partially named tensors

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#27308 Allow `align_to` to take in partially named tensors**

Currently, `tensor.align_to(*names)` has the restriction that the
`tensor` must be fully named. This doesn't need to be the case, when
using Ellipsis, we "expand the ellipsis to all unmentioned dimensions,
in the order which they appear in the original tensor".

For example, consider `tensor: Tensor[None, None, C]`.

`tensor.align_to(C, None, None)` is ambiguous because the user might
have wanted to switch the order of the None dimensions and there is no
way to specify that using this API. However, `tensor.align_to('C', ...)`
isn't ambiguous: we can select the two unnamed dimensions in the order
in which they appear.

To actually implement this, we write a brand-new `align_to(names,
ellipsis_idx)` function in c++ that is separate from the regular
`align_to(names)` implementation. Ideally we would support "..." as a
special name in c++ and combine the two implementations; we'll need to
support "..." in c++ in the future but that requires a bit of extra work.
In this PR, Python processees the ellipsis and then calls the correct
overload.

Test Plan:
- run tests

Differential Revision: [D17745179](https://our.internmc.facebook.com/intern/diff/D17745179)

zou3519 committed:
Allow `align_to` to take in partially named tensors

Currently, `tensor.align_to(*names)` has the restriction that the
`tensor` must be fully named. This doesn't need to be the case, when
using Ellipsis, we "expand the ellipsis to all unmentioned dimensions,
in the order which they appear in the original tensor".

For example, consider `tensor: Tensor[None, None, C]`.

`tensor.align_to(C, None, None)` is ambiguous because the user might
have wanted to switch the order of the None dimensions and there is no
way to specify that using this API. However, `tensor.align_to('C', ...)`
isn't ambiguous: we can select the two unnamed dimensions in the order
in which they appear.

To actually implement this, we write a brand-new `align_to(names,
ellipsis_idx)` function in c++ that is separate from the regular
`align_to(names)` implementation. Ideally we would support "..." as a
special name in c++ and combine the two implementations; we'll need to
support "..." in c++ in the future but that requires a bit of extra work.
In this PR, Python processees the ellipsis and then calls the correct
overload.

Test Plan:
- run tests

nairbv commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -69,6 +69,10 @@
   variants: method
   supports_named_tensor: True
 
+- func: align_to(Tensor(a) self, DimnameList order, int ellipsis_idx) -> Tensor(a)
```
this seems to me like an internal helper function that we shouldn't expose to users

zou3519 replied:
We're not exposing this to users. The Python `align_to` call wraps this function.

nairbv replied:
I thought every interface defined in native_functions.yaml becomes visible to users?

zou3519 replied:
In Python, torch.Tensor is a subclass of torch._C.TensorBase. The native functions interface generates align_to as a method of torch._C.TensorBase. I added an overwrite for `Tensor.align_to` in the Python definition of the class. Since users only interact with torch.Tensor (and not torch._C.TensorBase), they never see the native functions version of align_to.

zou3519 replied:
The above workflow I described is how you can override a method. We can also override torch functions as well (and we do so in cases where it is difficult to implement functionality in C++ for). One of these is [`broadcast_tensors`](https://github.com/pytorch/pytorch/blob/a6bb8b52d4cce3ea356354e88ce4d05659bb21de/torch/functional.py#L26-L52) : users cannot call the native_functions variant of it directly, they always to through the functional.py call.

zou3519 replied:
There's another long tail of things you can do in the C++ or codegen side to write functions in native_functions.yaml but have them not be directly exposed to Python. I can explain those in more detail if you're interested.

nairbv replied:
thanks, I think I would like to understand that better at some point. I had another case in a PR where I needed to exclude something from here to avoid exposing it.

zou3519 committed:
Update on "Allow `align_to` to take in partially named tensors"


Currently, `tensor.align_to(*names)` has the restriction that the
`tensor` must be fully named. This doesn't need to be the case, when
using Ellipsis, we "expand the ellipsis to all unmentioned dimensions,
in the order which they appear in the original tensor".

For example, consider `tensor: Tensor[None, None, C]`.

`tensor.align_to(C, None, None)` is ambiguous because the user might
have wanted to switch the order of the None dimensions and there is no
way to specify that using this API. However, `tensor.align_to('C', ...)`
isn't ambiguous: we can select the two unnamed dimensions in the order
in which they appear.

To actually implement this, we write a brand-new `align_to(names,
ellipsis_idx)` function in c++ that is separate from the regular
`align_to(names)` implementation. Ideally we would support "..." as a
special name in c++ and combine the two implementations; we'll need to
support "..." in c++ in the future but that requires a bit of extra work.
In this PR, Python processees the ellipsis and then calls the correct
overload.

Test Plan:
- run tests

Differential Revision: [D17745179](https://our.internmc.facebook.com/intern/diff/D17745179)
